leaf-server/src/main/java/org/dreeam/leaf/async/path/AsyncPathProcessor.java

```diff
@@ -106,27 +106,24 @@ private static BlockingQueue<Runnable> getQueueImpl() {
     private static @NotNull RejectedExecutionHandler getRejectedPolicy() {
         return (Runnable rejectedTask, ThreadPoolExecutor executor) -> {
             BlockingQueue<Runnable> workQueue = executor.getQueue();
-            if (!executor.isShutdown()) {
-                switch (AsyncPathfinding.asyncPathfindingRejectPolicy) {
-                    case FLUSH_ALL -> {
-                        if (!workQueue.isEmpty()) {
-                            List<Runnable> pendingTasks = new ArrayList<>(workQueue.size());
-
-                            workQueue.drainTo(pendingTasks);
-
-                            for (Runnable pendingTask : pendingTasks) {
-                                pendingTask.run();
-                            }
-                        }
-                        rejectedTask.run();
-                    }
-                    case CALLER_RUNS -> rejectedTask.run();
+
+            switch (AsyncPathfinding.asyncPathfindingRejectPolicy) {
+                case FLUSH_ALL -> {
+                    // Process in batches instead of all at once to avoid blocking main thread
+                    int batchSize = Math.min(32, workQueue.size());
+                    List<Runnable> batch = new ArrayList<>(batchSize);
+                    workQueue.drainTo(batch, batchSize);
+
+                    batch.forEach(Runnable::run);
+                    rejectedTask.run();
                 }
+                case CALLER_RUNS -> rejectedTask.run();
             }
 
-            if (System.currentTimeMillis() - lastWarnMillis > 30000L) {
-                LOGGER.warn("Async pathfinding processor is busy! Pathfinding tasks will be treated as policy defined in config. Increasing max-threads in Leaf config may help.");
-                lastWarnMillis = System.currentTimeMillis();
+            long currentTime = System.currentTimeMillis();
```

Comment:
Also just a nitpick: `currentTimeMillis` is not monotonic, `nanoTime` is. In other words, `currentTimeMillis` may cause code to misunderstand how much time elapsed when the system clock is changed, whereas for `nanoTime` it would work fine.